File: Data Structures & Algorithms/interleaving-string/submission-3.cpp
class Solution {
public:
    bool helper(string &s1, string &s2, string &s3, int i, int j , int k){
        if(i == s1.length() && k == s3.length() || j == s2.length() && k == s3.length()){
            return true;
        }
        bool ans = false;
        if(s1[i] == s3[k]){
            ans = ans | helper(s1,s2,s3,i+1,j,k+1);
        } 
        if(s2[j] == s3[k]) {
            ans = ans | helper(s1,s2,s3,i,j+1,k+1);
        }

        return ans;
    }
    bool isInterleave(string s1, string s2, string s3) {
        if(s1.length() + s2.length() != s3.length())
            return false;
        return helper(s1, s2, s3, 0, 0, 0);
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-3.java
class Solution {
    List<List<Integer>> s = new ArrayList<>();
    private void helper(int[] nums, int idx, List<Integer> l){
        if(idx >= nums.length){
            s.add(new ArrayList(l));
            return;
        }
        int j = idx;
        while((j < nums.length) && (nums[j] == nums[idx])){
            j++;
        }
        helper(nums, j, l);
        l.add(nums[idx]);
        helper(nums, idx+1, l);
        l.remove(l.size()-1);
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        helper(nums, 0, new ArrayList<>());
        return s.stream().collect(Collectors.toList());
    }
}
